[FIX] 로그인 시 카카오 앱으로 리다이렉트 안되는 현상 수정

stopstone opened the pull request:
## 작업 내용

### 1. 카카오 로그인 UI 계층으로 이동

- **LoginScreen에서 직접 카카오 로그인 처리**
  - Activity Context를 직접 사용하여 카카오톡 앱 전환 지원
  - 카카오 SDK 호출 로직을 presentation에 배치
  - 로그인 성공 시 ViewModel에 결과 전달

- **ViewModel은 비즈니스 로직만 처리**
  - 카카오 AccessToken을 받아 서버 전송 처리
  - UI 상태 관리 및 이벤트 처리
  - Repository를 통한 데이터 저장

- **불필요한 코드 제거**
  - Data Layer의 카카오 SDK 의존성 제거

### 2. 로그인 플로우 개선

- **토큰 전송 시점 변경**
  - 카카오 로그인 성공 → AccessToken만 임시 저장
  - 개인정보 동의 완료 → 서버 전송 및 DataStore 저장
  - 사용자 동의 전 서버 전송 방지

- **취소 처리 개선**
  - 카카오톡 앱에서 취소 시 로그인 취소로 처리
  - 웹뷰 fallback 제거하여 UX 개선

### 3. 카카오톡 앱 전환 기능

- **카카오톡 설치 여부에 따른 분기**
  - 카카오톡 설치 시 → 앱으로 이동하여 로그인
  - 카카오톡 미설치 시 → 웹뷰로 로그인
  - 자동으로 Near 앱으로 복귀


## 확인 방법

1. 카카오톡 설치 상태에서 로그인 버튼 클릭
2. 카카오톡 앱으로 이동하여 로그인
3. 로그인 성공 후 Near 앱으로 자동 복귀
4. 개인정보 동의 바텀시트에서 약관 동의 후 가입 버튼 클릭
5. 홈 화면으로 이동 확인

카카오톡 미설치 상태:
1. 로그인 버튼 클릭
2. 웹뷰로 로그인
3. 성공 후 Near 앱으로 복귀

## 참고 사항
- 한번이라고 로그인 된 계정이 있으면 카카오톡에 등록되어 리다이렉트 되지 않고, 바로 로그인 되는 현상이 있어
로그인 전에 한 번 unlink하고 있습니다!

## 관련 이슈
- close #49

stopstone committed:
feat: 카카오 소셜 로그인 로직 구현

- `LoginScreen`에서 카카오 SDK를 직접 호출하여 로그인하도록 로직을 구현했습니다.
- 카카오톡 앱 설치 여부에 따라 카카오톡 또는 카카오 계정으로 로그인을 시도하며, 실패 시 대체 로그인을 수행합니다.
- 로그인 성공 시 `ViewModel`의 `onSocialLoginSuccess`를, 실패 시 `onSocialLoginFailure`를 호출하도록 변경했습니다.
- `AuthRepository`에서 소셜 로그인 처리를 담당하던 `SocialLoginProcessor`를 제거하고, 외부에서 전달받은 액세스 토큰으로 서버 로그인을 요청하는 단일 메서드로 통합했습니다.

stopstone committed:
chore: 카카오 로그인 관련 Strategy 패턴 코드 삭제

- 소셜 로그인 처리를 위해 사용했던 Strategy 패턴 관련 코드들을 삭제했습니다.
- `SocialLoginDataSource` 인터페이스, `KakaoDataSource` 구현체, `SocialLoginProcessor`, 그리고 관련 `DataSourceModule`이 삭제되었습니다.

stopstone committed:
refactor: 카카오톡 로그인 실패 시 계정 로그인 재시도 로직 제거

- 사용자가 카카오톡 앱으로 로그인 시도 중 취소하는 경우를 포함하여, 로그인 실패 시 카카오 계정으로 재시도하던 로직을 제거했습니다.

stopstone committed:
refactor: 카카오 로그인 전 로그아웃 로직 추가

- 카카오 로그인(`loginWithKakao`) 함수 실행 시, 새로운 로그인을 보장하기 위해 기존 로그인 정보를 먼저 로그아웃하는 `logoutKakao()` 호출을 추가했습니다.

stopstone committed:
refactor: 소셜 로그인 로직을 개인정보 동의 후 실행하도록 변경

- 기존에 소셜 로그인 API를 호출하고 성공 시 개인정보처리방침 동의 바텀시트를 표시하던 방식에서, 바텀시트를 먼저 표시하도록 순서를 변경했습니다.
- 사용자가 개인정보처리방침에 동의했을 때(`onPrivacyConsentComplete` 호출 시) 소셜 로그인 API를 호출하도록 로직을 수정했습니다.

stopstone committed:
refactor: 카카오 로그인 시 로그아웃 대신 연결 끊기 호출

- 기존에 카카오 로그인 재시도 시 `logout()`을 호출하던 로직을 `unlink()` (연결 끊기)를 호출하도록 변경했습니다.
- 이에 따라 함수 이름도 `logoutKakao()`에서 `unlinkKakao()`로 수정했습니다.

stopstone committed:
refactor: ETC 로그인 `TODO()` 제거

- `LoginScreen`에서 `ProviderType.ETC` 분기에 있던 `TODO()` 호출 코드를 제거했습니다.

stopstone committed:
refactor: 카카오 로그인 시점에 회원 연결 끊기 수행

- 기존에 로그인 버튼 클릭 시 즉시 실행되던 `unlinkKakao()` 호출을, 실제 카카오 로그인을 수행하기 직전으로 이동시켰습니다.

stopstone committed:
refactor: 카카오 로그인 취소 시 예외 처리 로직 수정

- 사용자가 카카오 로그인을 취소했을 때, 빈 accessToken을 반환하는 대신 예외를 발생시키도록 로직을 수정했습니다.
- 로그인 취소 시 반환할 예외 메시지를 `strings.xml` 리소스(`login_user_cancelled`)로 추가하고 이를 사용하도록 변경했습니다.
- `loginWithKakaoTalk`과 `loginWithKakaoAccount`에서 `handleKakaoLoginResult`를 호출할 때, 로그인 취소 메시지를 전달하도록 수정했습니다.

stopstone committed:
feat: 소셜 로그인 시 액세스 토큰 유효성 검사 추가

- `AuthRepository`에서 소셜 로그인을 요청하기 전에 액세스 토큰이 비어있지 않고 최소 길이를 만족하는지 확인하는 `validateAccessToken` 로직을 추가했습니다.

stopstone committed:
refactor: ETC 로그인 타입 제거

- `ProviderType` enum에서 사용하지 않는 `ETC`를 제거했습니다.
- `LoginScreen`에서 `ETC` 타입 관련 분기 처리 로직을 제거했습니다.

stopstone committed:
feat: 카카오 로그인 실패 시 에러 케이스별 메시지 표시

- 카카오 로그인 실패 시 발생하는 에러를 `AuthError`, `ApiError`, `ClientError`, `KakaoSdkError` 등 구체적인 케이스로 나누어 처리하도록 구현했습니다.
- 각 에러 상황에 맞는 사용자 안내 메시지를 `strings.xml`에 추가하고, 로그인 실패 시 해당 메시지를 표시하도록 변경했습니다.

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/login/LoginViewModel.kt:
```diff
@@ -56,18 +56,26 @@ class LoginViewModel
+                _loginState.value.copy(
+                    socialLoginToken = accessToken,
+                    providerType = providerType,
+                    showPrivacyBottomSheet = true,
```
바텀시트를 보여준다는 것은 view의 의존성이 있어보입니다!
동의를 받아야한다 등 어떤 목적인지에 대한 네이밍으로 고민해보면 어떨까요?

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/login/LoginViewModel.kt:
```diff
@@ -76,16 +84,44 @@ class LoginViewModel
+                        updateLoadingState(isLoading = false)
+                        _loginState.value =
+                            _loginState.value.copy(
+                                showPrivacyBottomSheet = false,
```
여기도 위 코멘트와 같이 확인 부탁드립니다!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/login/LoginScreen.kt:
```diff
@@ -212,6 +243,102 @@ private object LoginScreenConstants {
+                Result.failure(Exception(cancelledMessage)),
+            )
+    }
+}
```
Screen이 로직을 가져오면서 무거워진 것 같습니다😅
이 로직들은 별도의 클래스로 분리하여 remember로 관리하는 것은 어떨까요?

stopstone replied:
감사합니다!!
로그인 로직을 분리한다면.. 패키지 네이밍과 위치는 어떻게 하면 좋을까요?

현재 생각하는 것은 presentation/feature/login/auth/KakaoLogin~.kt 로 하면 좋을 것 같다고 생각합니다!

rhkrwngud445 replied:
저는 멀티모듈 기준 최상단 또는 core내부가 적절해 보이는데요!
현재는 해당 login 패키지내 구성하고 이슈나 todo로 리마인드 될 수 있게 하면 좋을 것 같습니다!

현재 말씀주신 방법으로 구현주시면 될 것 같아요!

stopstone replied:
코멘트 감사합니다!
우선은 login 패키지에 두고 이후 core로 이동시키겠습니다!

core는 공통, 재사용 가능한 것들을 넣는다는 생각에 생각 못했는데 감사합니다!

rhkrwngud445 reviewed (commented):
큰 수정을 이렇게.. 감사합니다!👍
가독성을 개선해볼만한 코멘트로 조금 남겼습니다! 확인 부탁드려요!

고생하셨습니다!😀

stopstone committed:
refactor: 소셜 로그인 로직 추상화

- 각 소셜 로그인 구현을 통합적으로 관리하기 위해 `SocialLoginHandler`를 도입했습니다.
- 다양한 소셜 로그인 방식의 공통 인터페이스로 `SocialLoginProvider`를 정의했습니다.
- 향후 새로운 소셜 로그인(Apple, Google 등)을 쉽게 추가할 수 있는 확장 가능한 구조를 마련했습니다.

stopstone committed:
feat: 카카오 로그인 기능 구현

- 카카오 SDK를 사용하여 소셜 로그인을 처리하는 `KakaoLoginManager` 클래스를 추가했습니다.
- 카카오톡 앱 설치 여부에 따라 카카오톡 또는 카카오 계정으로 로그인을 분기하여 처리합니다.

stopstone committed:
refactor: 소셜 로그인 로직 분리

- `LoginScreen.kt`에 있던 카카오 로그인 관련 로직을 `SocialLoginHandler` 클래스로 분리했습니다.
- `LoginRoute`에서 `SocialLoginHandler`를 생성하고 사용하여 로그인 로직을 호출하도록 수정했습니다.

stopstone committed:
refactor: 'showPrivacyBottomSheet' 프로퍼티명을 'requiresPrivacyConsent'로 수정

stopstone commented:
@rhkrwngud445 리뷰 감사합니다 리뷰주신 것을 반영해봤습니다!
로그인 로직을 분리하는 김에 나중에 소셜로그인이 추가될 가능성을 고려하여 SocialLoginProvider(인터페이스), SocialLoginHandler(구현체)를 추가하였습니다 !

실제 카카오 로그인 관련은 KakaoLoginManager에 구현하였고, 모두 feature/login/auth에 위치하였습니다 :)

rhkrwngud445 commented:
잘 추상화된 것 같네요! 바로 머지 주셔도 될 것 같습니다👍
고생하셨습니다!
